### PR TITLE
Fix definite assignment analysis of local function out parameters

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Fixes https://github.com/dotnet/roslyn/issues/41631

cc @RikkiGibson 

I did not address https://github.com/dotnet/roslyn/pull/41597#issuecomment-585488435 since I don't believe the issue is reproducible until https://github.com/dotnet/roslyn/issues/41898 is fixed.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs [hunk 1]
```diff
@@ -1550,5 +1550,49 @@ System.Collections.Generic.IEnumerable<string> getGoo(int count, out bool output
...
+public class C
+{
+    void M()
+    {
+        string s0;
+        local1(out s0);
+    
+        void local1(out string s1)
+        {
+            s0.ToString(); // should produce an error, but doesn't
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs]: please remove the comment :)

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 2]
```diff
@@ -1294,6 +1298,12 @@ public override BoundNode VisitEventAssignmentOperator(BoundEventAssignmentOpera
         }
 
         protected virtual void VisitArguments(ImmutableArray<BoundExpression> arguments, ImmutableArray<RefKind> refKindsOpt, MethodSymbol method)
+        {
+            VisitArgumentsBeforeCall(arguments, refKindsOpt);
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: Is this method just for non-method calls that take arguments? indexers, etc.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 3]
```diff
@@ -6,6 +6,7 @@
 using System.Collections.Immutable;
 using System.Diagnostics;
 using System.Text;
+using Microsoft.CodeAnalysis;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: did the code change require this additional using?

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs (hunk 2 above)
Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: It's also used by indexers. I don't see why that matters though?

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs (hunk 1 above)
Comment (YairHalberstadt) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs]: I haven't seen `// 1` before. What's it for?

**Comment by RikkiGibson** [comment]
@dotnet/roslyn-compiler could we please get a senior review on a small bug fix?

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs (hunk 2 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: just checking my understanding :)

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs]: it's a convention that makes it easier to tell at a glance which lines are expected to produce diagnostics. So if your test produced several diagnostics you may choose to annotate some lines with `// 1`, `// 2`, etc.

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 4]
```diff
@@ -1294,6 +1297,12 @@ public override BoundNode VisitEventAssignmentOperator(BoundEventAssignmentOpera
         }
 
         protected virtual void VisitArguments(ImmutableArray<BoundExpression> arguments, ImmutableArray<RefKind> refKindsOpt, MethodSymbol method)
+        {
+            VisitArgumentsBeforeCall(arguments, refKindsOpt);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: >VisitArgumentsBeforeCall(arguments, refKindsOpt); [](start = 12, length = 49)

Please add an assert that method is not a local function. It also would be good to add a doc comment about that requirement.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs [hunk 5]
```diff
@@ -1550,5 +1550,49 @@ System.Collections.Generic.IEnumerable<string> getGoo(int count, out bool output
...
+        [WorkItem(41631, "https://github.com/dotnet/roslyn/issues/41631")]
+        public void UseOfCapturedVariableAssignedByOutParameter()
+        {
+            CreateCompilation(@"
+public class C
+{
+    void M()
+    {
+        string s0;
+        local1(out s0); // 1
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs]: Could you also add a test like

```C#
void M()
{
    string s0;
    local1(s0 = ""hello"", out s0);

    void local1(string s1, out string s2)
    {
           s0.ToString();
           s2 = ""bye"";
    }
}
```

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 6]
```diff
@@ -1296,8 +1296,12 @@ public override BoundNode VisitEventAssignmentOperator(BoundEventAssignmentOpera
             return null;
         }
 
+        /// <summary>
+        /// Do not call for a local function.
+        /// </summary>
         protected virtual void VisitArguments(ImmutableArray<BoundExpression> arguments, ImmutableArray<RefKind> refKindsOpt, MethodSymbol method)
         {
+            Debug.Assert(!(method.OriginalDefinition is LocalFunctionSymbol));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: 'method' can be null here, so consider using a condition like `method?.OriginalDefinition.MethodKind != MethodKind.LocalFunction`

Also, is there any situation where some constructed method's MethodKind differs from the OriginalDefinition's MethodKind? It feels like we may be able to just check `method?.MethodKind`?

Comment (YairHalberstadt) [reply src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs]: We check the OriginalDefinition here: https://github.com/dotnet/roslyn/pull/41899/files#diff-35d4070efa937f088ceaea2864b3279fL1165

I presume that's for a reason.

**Pull request merged**